File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0,
            r = height.size() - 1,
            volMax = 0;
        
        while (l < r) {
            int vol = min(height[l],height[r]) * (r - l);
            volMax = max(vol,volMax);
            
            if (height[l] < height[r]) {
                ++l;
            } else {
                --r;
            }
        }
        
        return volMax;
    }
};

File: 0011-container-with-most-water/0011-container-with-most-water.rs
use std::cmp::Ordering;

impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let n = height.len();
        
        if n < 2 {
            /// insufficient length for height vector
            return 0;
        }
        
        let (mut left, mut right, mut volMax) = (0, height.len() - 1, 0);
        
        while left < right {
            let vol = height[left].min(height[right]) * ((right - left) as i32);
            volMax = volMax.max(vol);
            
            if height[left] > height[right] {
                right -= 1;
            } else {
                left += 1;
            }
        }
        
        volMax
    }
}